ddl/ddl_api.go

```diff
@@ -5245,66 +5193,106 @@ func checkPlacementSpecConstraint(rules []*placement.Rule, rule *placement.Rule,
 		}
 
+		ruleCnt := int(replicas)
 		for labels, cnt := range constraints {
-			newRule := &placement.Rule{}
-			*newRule = *rule
+			newRule := rule.Clone()
 			if cnt <= 0 {
-				err = errors.Errorf("count should be non-positive, but got %d", cnt)
+				err = errors.Errorf("count should be positive, but got %d", cnt)
 				break
 			}
-			// TODO: handle or remove rule.Count in later commits
-			rule.Count -= cnt
+			if replicas != 0 {
+				ruleCnt -= cnt
+				if ruleCnt < 0 {
+					err = errors.Errorf("REPLICAS should be larger or equal to the number of total replicas, but got %d", replicas)
+					break
+				}
+			}
 			newRule.Count = cnt
-			err = checkPlacementLabelConstraints(newRule, strings.Split(strings.TrimSpace(labels), ","))
+
+			newRule.LabelConstraints, err = placement.CheckLabelConstraints(strings.Split(strings.TrimSpace(labels), ","))
 			if err != nil {
 				break
 			}
 			rules = append(rules, newRule)
 		}
+		rule.Count = ruleCnt
+
+		if rule.Count > 0 {
+			rules = append(rules, rule)
+		}
 	} else {
 		err = errors.Errorf("constraint should be a JSON array or object, but got '%s'", cnstr)
 	}
 	return rules, err
 }
 
-func checkPlacementSpecs(specs []*ast.PlacementSpec) ([]*placement.Rule, error) {
-	rules := make([]*placement.Rule, 0, len(specs))
+func buildPlacementSpecs(specs []*ast.PlacementSpec, tableID, partitionID int64) ([]*placement.RuleOp, error) {
+	rules := make([]*placement.RuleOp, 0, len(specs))
 
 	var err error
 	var sb strings.Builder
 	restoreFlags := format.RestoreStringSingleQuotes | format.RestoreKeyWordLowercase | format.RestoreNameBackQuotes
 	restoreCtx := format.NewRestoreCtx(restoreFlags, &sb)
 
 	for _, spec := range specs {
-		rule := &placement.Rule{
-			GroupID:  placementRuleDefaultGroupID,
-			Count:    int(spec.Replicas),
-			Override: true,
+		rule := &placement.RuleOp{
+			Rule: &placement.Rule{
+				GroupID:  placement.RuleDefaultGroupID,
+				Override: true,
+			},
 		}
 
-		switch spec.Tp {
-		case ast.PlacementAdd:
+		switch spec.Role {
+		case ast.PlacementRoleFollower:
+			rule.Role = placement.Follower
+		case ast.PlacementRoleLeader:
+			rule.Role = placement.Leader
+		case ast.PlacementRoleLearner:
+			rule.Role = placement.Learner
+		case ast.PlacementRoleVoter:
+			rule.Role = placement.Voter
 		default:
-			err = errors.Errorf("unknown action type: %d", spec.Tp)
+			err = errors.Errorf("unknown role: %d", spec.Role)
 		}
 
 		if err == nil {
-			switch spec.Role {
-			case ast.PlacementRoleFollower:
-				rule.Role = placement.Follower
-			case ast.PlacementRoleLeader:
-				rule.Role = placement.Leader
-			case ast.PlacementRoleLearner:
-				rule.Role = placement.Learner
-			case ast.PlacementRoleVoter:
-				rule.Role = placement.Voter
+			switch spec.Tp {
+			case ast.PlacementAdd:
+				rule.Action = placement.RuleOpAdd
+			case ast.PlacementAlter:
+				rule.Action = placement.RuleOpAdd
+
+				// alter will overwrite all things
+				// drop all rules that will be overrided
```

Comment:
```suggestion
				// drop all rules that will be overridden
```
Is it necessary to throw a warning if a `spec` overrides the previous `specs`?

Reply:
No. `Alter` is intended to override all related rules.